File: tags/2.0-M3/device-agent-simulator/src/main/java/org/openmobster/device/agent/sync/SyncService.java
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.openmobster.device.agent.sync;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.Vector;

import org.openmobster.core.common.IOUtilities;
import org.openmobster.device.agent.Tools;
import org.openmobster.device.agent.sync.engine.SyncEngine;
import org.openmobster.device.agent.sync.engine.ChangeLogEntry;
import org.openmobster.device.agent.configuration.Configuration;

/**
 * @author openmobster@gmail.com
 */
public class SyncService 
{
	private SyncEngine syncEngine = null;
	
	public static String TWO_WAY = "200";
	public static String SLOW_SYNC = "201";
	public static String ONE_WAY_CLIENT = "202";
	public static String ONE_WAY_SERVER = "204";
	
	public void start()
	{
		
	}
	
	public void stop()
	{
		
	}
	
	public void setSyncEngine(SyncEngine syncEngine)
	{
		this.syncEngine = syncEngine;
	}
	
	public SyncEngine getSyncEngine()
	{
		return this.syncEngine;
	}
	//---------------------------------------------------------------------------------------------------------
	public void startSync(String syncType, String deviceId, String serverId, String deviceService, 
	String serverService)
	{
		try
		{
			Socket socket = null;
			OutputStream os = null;
			InputStream is = null;
			try
			{					
				socket = Tools.getPlainSocket();
				
				is = socket.getInputStream();
				os = socket.getOutputStream();	
								
				String authHash = Configuration.getInstance().getAuthenticationHash();				
				String sessionInitPayload = "<auth>"+deviceId+"|"+authHash
				+"</auth>&processorid=sync";
				
				IOUtilities.writePayLoad(sessionInitPayload, os);
				
				String data = IOUtilities.readServerResponse(is);
				if(data.indexOf("status=200")!=-1)
				{
					this.performSync(syncType, is, os, data, 5000,
				    deviceId, serverId, deviceService, serverService);
				}
			}
			finally
			{					
				if(socket != null)
				{
					socket.close();
				}
			}
		}
		catch(Exception e)
		{
			throw new RuntimeException(e);
		}
	}
	
	public void updateChangeLog(String service, String operation, String recordId)
	throws SyncException
	{
		Vector changelog = new Vector();
		ChangeLogEntry entry = new ChangeLogEntry();
		entry.setNodeId(service);
		entry.setOperation(operation);
		entry.setRecordId(recordId);
		changelog.add(entry);
		
		this.syncEngine.addChangeLogEntries(changelog);
	}
	//------------------------------------------------------------------------------------------------------------
	private void performSync(String syncType, InputStream is, OutputStream os, String data,
	int maxClientSize,String deviceId, String serverId, String deviceService, 
	String serverService)
	throws SyncException, IOException
	{	
		//Get the Client Simulator
		SyncAdapter clientSyncAdapter = new SyncAdapter();
		clientSyncAdapter.setSyncEngine(syncEngine);
		
    	//Get the initialization payload
		SyncAdapterRequest request = new SyncAdapterRequest();
		request.setAttribute(SyncAdapter.SOURCE,deviceId);
		request.setAttribute(SyncAdapter.TARGET, serverId);
		request.setAttribute(SyncAdapter.DATA_SOURCE, deviceService);
		request.setAttribute(SyncAdapter.DATA_TARGET, serverService);
		request.setAttribute(SyncAdapter.MAX_CLIENT_SIZE, String.valueOf(maxClientSize));
		request.setAttribute(SyncAdapter.CLIENT_INITIATED, "true");		
		request.setAttribute(SyncAdapter.SYNC_TYPE, syncType);
		
		//Start the synchronization session
		SyncAdapterResponse response = clientSyncAdapter.start(request);
		
		//Setup the payload to be sent to the server
		String payLoad = (String)response.getAttribute(SyncAdapter.PAYLOAD);		
		
		//Start sync session
		IOUtilities.writePayLoad(payLoad, os);
																			
		data = IOUtilities.readServerResponse(is);
		
		
		
		//Orchestrate the synchronization session until it is successfully finished
		while(true)
		{
			request = new SyncAdapterRequest();
			request.setAttribute(SyncAdapter.PAYLOAD, data);
			response = clientSyncAdapter.service(request);			
			
			payLoad = (String)response.getAttribute(SyncAdapter.PAYLOAD);			
						
			if(response.getStatus() == SyncAdapter.RESPONSE_CLOSE)
			{
				break;
			}
			
						
			//Send the payload to the server for processing			
			IOUtilities.writePayLoad(payLoad, os);
			
			data = IOUtilities.readServerResponse(is);				
		}
	}			
}

File: trunk/iphone/mobilecloudlib/SyncService.m
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#import "SyncService.h"


/**
 * FIXME: Implement proper synchronization strategy when Channel Locking is figured
 * out
 * 
 * @author openmobster@gmail.com
 */
@implementation SyncService

-init
{
	if(self == [super init])
	{
		engine = [[SyncEngine alloc] init];
	}
	
	return self;
}

-(void)dealloc
{
	[engine release];
	[super dealloc];
}

+(SyncService *) getInstance
{
	Registry *registry = [Registry getInstance];
	return (SyncService *)[registry lookup:[SyncService class]];
}

-(void)startDaemons
{
     /*NSArray *changelog = [engine getChangeLog:@"phonegap_channel" operation:@"Replace"];
     NSLog(@"-----ChangeLog Dump On the Main Thread-----------------------");
     if(changelog != nil && [changelog count]>0)
     {
     for(ChangeLogEntry *local in changelog)
     {
     NSString * recordId = local.recordId;
     NSLog(@"RecordId(Main): %@",local.recordId);
     NSLog(@"--------------------------------------");
     }
     }
     else 
     {
     NSLog(@"ChangeLog is Empty!!!");
     }*/
    
    //persist the managedcontext in the database
    /*NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
	
	NSError* error;
	BOOL success = [managedContext save:&error];
    if(!success)
    {
        NSLog(@"Error during save!!!");
        NSLog(@"SaveError: %@, %@", error, [error userInfo]);
    }
    
    NSLog(@"Save Success:%d",success);*/
    
	//Schedule a background sync to sync up the channes at App startup
	SyncScheduler *backgroudSync = [SyncScheduler getInstance];
	[backgroudSync startBackgroundSync];
}

-(void)startDeltaSync:(NSString *)channel
{
    SyncScheduler *backgroudSync = [SyncScheduler getInstance];
	[backgroudSync startDeltaSync:channel];
}

-(void)performBootSync:(NSString *)channel :(BOOL)isBackground
{
	@synchronized(self)
	{
		[self startSync:_BOOT_SYNC :channel :isBackground :nil];
	}
}

-(void)performTwoWaySync:(NSString *)channel :(BOOL)isBackground
{
	@synchronized(self)
	{
		[self startSync:_TWO_WAY :channel :isBackground :nil];
	}
}

-(void)performOneWayServerSync:(NSString *)channel :(BOOL)isBackground
{
	@synchronized(self)
	{
		[self startSync:_ONE_WAY_SERVER :channel :isBackground :nil];
	}
}

-(void)performOneWayClientSync:(NSString *)channel :(BOOL)isBackground
{
	@synchronized(self)
	{
		[self startSync:_ONE_WAY_CLIENT :channel :isBackground :nil];
	}
}

-(void)performSlowSync:(NSString *)channel :(BOOL)isBackground
{
	@synchronized(self)
	{
		[self startSync:_SLOW_SYNC :channel :isBackground :nil];
	}
}

-(void)performStreamSync:(NSString *)channel :(BOOL)isBackground :(NSString *)oid
{
	@synchronized(self)
	{
		[self startSync:_STREAM :channel :isBackground :oid];
	}
}

-(void)updateChangeLog:(NSString *)channel :(NSString *) operation :(NSString *) objectId
{
	@synchronized(self)
	{
		NSMutableArray *entries = [NSMutableArray array];
	
		GenericAttributeManager *attr = [GenericAttributeManager withInit];
		[attr setAttribute:@"nodeId" :channel];
		[attr setAttribute:@"operation" :operation];
		[attr setAttribute:@"recordId" :objectId];
	
		[entries addObject:attr];
	
		[engine addChangeLogEntries:entries];
        
        
		//sync in the background
		[self startDeltaSync:channel];
	}
}


//Used internally
-(void)performSync:(NetSession *)session :(NSString *)syncType :(NSString *)channel :(BOOL)isBackground :(NSString *)oid
{
	Configuration *conf = [Configuration getInstance];
	SyncAdapter *syncAdapter = [SyncAdapter withInit];
	
	//Setup the Sync Initialization payload
	SyncAdapterRequest *request = [SyncAdapterRequest withInit];
	[request setAttribute:_SOURCE :conf.deviceId];
	[request setAttribute:_TARGET :conf.serverId];
	[request setAttribute:_MAX_CLIENT_SIZE :conf.maxPacketSize];
	[request setAttribute:_CLIENT_INITIATED :[NSNumber numberWithBool:YES]];
	[request setAttribute:_DATA_SOURCE :channel];
	[request setAttribute:_DATA_TARGET :channel];
	[request setAttribute:_SYNC_TYPE :syncType];
	[request setAttribute:@"isBackgroundSync" :[NSNumber numberWithBool:isBackground]];
	if(![StringUtil isEmpty:oid])
	{
		[request setAttribute:_STREAM_RECORD_ID :oid];
	}
    
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    NSString *appId = [bundle bundleIdentifier];
    [request setAttribute:_App :appId];
	
	//Start the Sync Session
	SyncAdapterResponse *response = [syncAdapter service:request];
	
	//Payload to start the sync session
	NSString *devicePayload = (NSString *)[response getAttribute:_PAYLOAD];
	
	//Sending the start request
	NSString *cloudPayload = [session sendPayload:devicePayload];
	
	//SyncSession orchestration
	while(YES)
	{
		request = [SyncAdapterRequest withInit];
		[request setAttribute:_PAYLOAD :cloudPayload];
		
		response = [syncAdapter service:request];
		devicePayload = (NSString *)[response getAttribute:_PAYLOAD];
		
		//Check for end of sync session
		if(response.status == _RESPONSE_CLOSE)
		{
			//Sync Session successfully ended
			break;
		}
		
		cloudPayload = [session sendPayload:devicePayload];
		if([cloudPayload isEqualToString:@"status=500"])
		{
			//Unknown CloudSide error occurred
			break;
		}
	}
}

-(void)startSync:(NSString *)syncType :(NSString *)channel :(BOOL)isBackground :(NSString *)oid
{
	NetSession *session = NULL;
	@try
	{
		Configuration *conf = [Configuration getInstance];
		BOOL secure = [conf.sslActive boolValue];
		NetworkConnector *connector = [NetworkConnector getInstance];
		session = [connector openSession:secure];
		
		NSString *handshake = NULL;
		NSString *deviceId = conf.deviceId;
		NSString *authHash = conf.authenticationHash;
		handshake = [NSString stringWithFormat:_SYNC_PAYLOAD,deviceId,authHash];
		
		
		NSString *response = [session performHandshake:handshake];
		if([StringUtil indexOf:response :@"status=200"] != -1)
		{
			[self performSync:session :syncType :channel :isBackground :oid];
		}
        else
        {
            SystemException *syse = [SystemException withContext:@"Start Sync" method:@"startSync" parameters:[NSMutableArray arrayWithObjects:response,nil]];
            @throw syse;
        }
	}
	@finally 
	{
		if(session != NULL)
		{
			[session close];
		}
	}
}
@end
